### PR TITLE
Use conditional conformances to implement Equatable for Optional, Array and Dictionary

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce conditional conformances to `Equatable` for `Optional`, `Array`, and `Dictionary`.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please benchmark

File: stdlib/public/core/Arrays.swift.gyb [hunk 1]
```diff
@@ -2181,57 +2181,55 @@ extension _ArrayBufferProtocol {
...
 
-/// Returns `true` if the arrays do not contain the same elements.
-@_inlineable
-public func != <Element : Equatable>(
-  lhs: ${Self}<Element>, rhs: ${Self}<Element>
-) -> Bool {
-  return !(lhs == rhs)
+  /// Returns `true` if the arrays do not contain the same elements.
+  @_inlineable
+  public static func !=(lhs: ${Self}<Element>, rhs: ${Self}<Element>) -> Bool {
```
Comment (xwu) [inline stdlib/public/core/Arrays.swift.gyb]: This shouldn't be necessary, now that we're conforming to `Equatable`, right?

File: stdlib/public/core/Optional.swift [hunk 2]
```diff
@@ -313,98 +313,100 @@ func _diagnoseUnexpectedNilOptional(_filenameStart: Builtin.RawPointer,
...
+  ///     if numberToFind != numberFromString {
+  ///         print("No match.")
+  ///     }
+  ///     // Prints "No match."
+  ///
+  /// - Parameters:
+  ///   - lhs: An optional value to compare.
+  ///   - rhs: Another optional value to compare.
+  @_inlineable
+  public static func !=(lhs: Wrapped?, rhs: Wrapped?) -> Bool {
```
Comment (xwu) [inline stdlib/public/core/Optional.swift]: Ditto here.

**Comment by shahmishal** [comment]
ci.swift.org is down for update, I will re-trigger this PR once update has been complete

**Comment by shahmishal** [comment]
@swift-ci please benchmark

**Comment by shahmishal** [comment]
@swift-ci please smoke test

**Comment by shahmishal** [comment]
@swift-ci please smoke test

**Comment by shahmishal** [comment]
@swift-ci please benchmark

**Comment by DougGregor** [comment]
Thank you @shahmishal !

File: stdlib/public/core/Arrays.swift.gyb (hunk 1 above)
Comment (DougGregor) [reply stdlib/public/core/Arrays.swift.gyb]: Currently crashes the compiler; I'll handle it as a follow-up, I think.

**Comment by DougGregor** [comment]
Some regressions, slightly more improvements. There are some inlining changes that I don't understand and still need to look into.

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by moiseev** [comment]
OMG! This is happening! 🍾

**Comment by gparker42** [comment]
Did this change break test SIL/parse_stdlib.sil ?
```
/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/validation-test-linux-x86_64/SIL/Output/parse_stdlib.sil.tmp.sil:1286711:28: error: sil associated type path for conformance not found τ_0_0
  conditional_conformance (τ_0_0: Equatable): dependent
```
You only ran smoke tests which doesn't run that test, I think.

**Comment by moiseev** [comment]
FWIW, this test is marked as `long_test` so it is not run even in full CI PR testing. It only runs on nightly CI builds, or something like that.

**Comment by gparker42** [comment]
That can't be right because the test failure broke my PR test.

**Comment by gparker42** [comment]
Still broken. I'm reverting in #12949.

**Comment by DougGregor** [comment]
Thanks, @gparker42.

**Comment by DougGregor** [comment]
For anyone watching from afar... this got reverted, and we realized that we should really go through the swift-evolution process before landing the standard library changes here. So, we'll be back with a Bigger and Better pull request as we figure out how to apply conditional conformances to the standard library.

**Comment by xwu** [comment]
What's the rationale for going through Swift Evolution for this?

**Comment by DougGregor** [comment]
It's a change to the standard library (an addition) and part of a larger effort to introduce conditional conformances into the standard library.

**Comment by xwu** [comment]
These changes were explicitly described in SE-0143, under "Introduction," "Motivation," and "Effect on ABI Stability." Having been through two rounds of review (in addition to commentary via the _Generics Manifesto_ threads), it's been thoroughly vetted, has it not?

**Comment by DougGregor** [comment]
Well, motivation isn't really a proposal, and there's a bunch of other changes we'd like to make in the standard library using conditional conformances that definitely need review. I'm re-opening this over at https://github.com/apple/swift/pull/13046 and will ask the core team for guidance... I'd really like to get it in.

**Comment by frogcjn** [comment]
will you bring [Extending protocols to conform to protocols](https://github.com/apple/swift-evolution/blob/master/proposals/0143-conditional-conformances.md#extending-protocols-to-conform-to-protocols) and [Overloading across constrained extensions](https://github.com/apple/swift-evolution/blob/master/proposals/0143-conditional-conformances.md#overloading-across-constrained-extensions)?

**Comment by DougGregor** [comment]
Neither of those have a design or a proposal, so no. Conditional conformances is already a large feature as it is.

**Pull request merged**